src/tensor_ops/utilities/reduction_utils.rs

```diff
@@ -1,8 +1,48 @@
-#[cfg(feature = "cuda")]
-use crate::prelude::{Axes, Shape};
+use crate::shapes::{Axes, Shape};
+use crate::tensor::cpu::NdIndex;
 #[cfg(feature = "cuda")]
 use std::vec::Vec;
 
+/// Permutes strides so that all reduces axes are rightmost.
+/// For example reducing (2, 3, 4) to (4,) would permute the shape to (4, 2, 3)
+/// via the strides.
+///
+/// This makes all the reduced elements sequential when indexed using [NdIndex].
+#[inline(always)]
+pub(crate) fn index_for_reductions<S: Shape, Ax: Axes>(
```

Comment:
@nkoppel i'm pretty sure this does exactly what the cuda version below does. Thoughts on if we could combine them? I'd like to keep the cpu one not using vecs, which I think is the biggest difference.

Reply:
I think we can factor out the shared logic from index_for_reductions and permute_for_reductions, but I'd prefer to keep the filtering of broadcasted dimensions in permute_for_reductions.

This implementation looks correct to me, and is likely faster than my implementation, so go ahead and use yours.